pkg/controller/nodes/executor.go

```diff
@@ -461,6 +472,59 @@ func (c *nodeExecutor) RecursiveNodeHandler(ctx context.Context, w v1alpha1.Exec
 	return executors.NodeStatusUndefined, errors.Errorf(errors.IllegalStateError, currentNode.GetID(), "Should never reach here")
 }
 
+func (c *nodeExecutor) FinalizeHandler(ctx context.Context, w v1alpha1.ExecutableWorkflow, currentNode v1alpha1.ExecutableNode) error {
+	nodeStatus := w.GetNodeExecutionStatus(currentNode.GetID())
+	switch nodeStatus.GetPhase() {
+	case v1alpha1.NodePhaseRunning, v1alpha1.NodePhaseFailing, v1alpha1.NodePhaseSucceeding, v1alpha1.NodePhaseRetryableFailure, v1alpha1.NodePhaseQueued:
```

Comment:
I think we should be finalizing only if it is succeeding, failing or retryable failure.
Abort automatically finalizes?